### PR TITLE
Resolve contract keys conflicts in disclosed contracts

### PR DESCRIPTION
**Pull request by mziolekda**
### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTableH2.scala
Comment (gerolf-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTableH2.scala:30]:
```suggestion
    keyNullifies = buildNullifyPastKeys(info),
```

**Review by gerolf-da — APPROVED** [review]
Thanks, @mziolekda. I'd like somebody else from the team to also review this before merging.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala
Comment (cocreature) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala:26]: I think this could benefit from some comments somewhere on why this is a safe thing to do.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala
Comment (oliverse-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala:27]: Is there a possibility that nulling out key hashes may break multi-key semantics in canton? cc @andreaslochbihler-da

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionIndexing.scala
Comment (cocreature) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionIndexing.scala:212]: This seems slightly off. I think it makes more sense to serialize later like we do for everything else afaict.

**Review by cocreature — APPROVED** [review]
Nice work, thanks!

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala
Comment (meiersi-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala:27]: Yes, I'm pretty sure that this change breaks multi-key semantics, as there it does not hold that seeing a create with a key means that all previous creates with that key must have been archived.

That said: this change is for the legacy IndexDB schema only. In the append-only schema it is not required.

Comment (cocreature) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala:27]: We don’t make any promises though which contract you can fetch in multi-key semantics. So it’s not entirely clear to me that this actually breaks anything.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala
Comment (meiersi-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala:85]:
```suggestion
      nullifyPastKeys: Option[BatchSql],
```
consistency with the other fields and the naming in other parts of the code

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionIndexing.scala
Comment (meiersi-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionIndexing.scala:212]:
```suggestion
      val netCreateKeys = netCreates.flatMap(create =>
```

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionsWriter.scala
Comment (meiersi-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionsWriter.scala:53]: Metrics are missing

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala
Comment (oliverse-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala:27]: But in the following scenario:

1. Create contract1 with key key1
2. Create contract2 with key key1 (nulls out contract1's key1 reference)
3. Archive contract2
4. FetchByKey(key1) would fail to look up key1

The canton tests still run on the legacy index. If we had an indexer configuration to opt out the nulling out, we could use this in canton.

**Review by meiersi-da — COMMENTED** [review]
Thanks for creating this patch @mziolekda . It looks functionally correct.

I've added a bunch of comments for simple code cleanup, but don't consider them must-haves for the patch, if we're really short on time.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala
Comment (andreaslochbihler-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala:27]: Indeed, this changes the behavior. I agree with @cocreature that this is probably still within the guarantees for "fetchSome" semantics. But the UX will degrade with this change even on a single non-UCK domains. Example topology: Alice (key maintainer) on P1, Bob on P2.

1. Alice creates a contract with key and observer Bob. Bob concurrently creates another contract with the same key. Due to concurrency, both transactions are accepted. 
2. Bob archives his contract.
3. Alice tries to lookup the key and finds nothing, even though her contract is still active (and reported as active by the ACS service).

I haven't tried it out but I suspect that this scenario used to work in Canton.

Comment (cocreature) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala:27]: I think it’s somewhat debatable if UX degrades or not. If you do step 2 and 3 in a single transaction you get `Nothing` at the moment, if you do it over multiple transactions you get a key (assuming your suspicion is correct). It’s not clear to me whether changing the multi-transaction behavior to match how things work within a single transaction is really worse.
More generally, I think this gets to my general point that multi-key semantics are ill-defined at the moment and I think worrying about UX without even being clear on what exactly the current semantics are is not all that productive.

Comment (andreaslochbihler-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala:27]: How would you be able to merge steps 2 and 3? They are submitted over distinct participants.

Comment (cocreature) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala:27]: I can run a submission as Alice that gets authority from Bob somehow to archive the contract. And more generally, I think we shouldn’t make the behavior here topology dependent.

**Review by tudor-da — APPROVED** [review]

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala
Comment (andreaslochbihler-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala:27]: > And more generally, I think we shouldn’t make the behavior here topology dependent.

I agree that the *allowed* behaviors should not depend on the topology. But the *observed* behavior does depend on it, because different topologies lead to different partial knowledge.

Comment (daravep) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala:27]: @cocreature @bame-da @meiersi-da @mziolekda @gerolf-da  @vivd-da @sormeter 

I've discussed this with @oliverse-da and @andreaslochbihler-da.

1. This patch will very much likely break current Canton on old schema running with non-UCK mode.
2. We acknowledge the urgency of this and ask you to not hesitate to proceed with his on the old schema. 
3. We assume that Canton GA and DDF4O for HKEX will be on the parallel indexer schema where this is not an issue.
4. Until we have fully migrated to the new indexer schema, we will mark all upcoming Canton releases as "DEFECT IN CERTAIN CASES" (June, July).
5. We will just disable any test that fails as a result of this change for now.
6. @oliverse-da will first finish his work on adding DAML exceptions to Canton before enabling the parallel indexer schema as default.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTableH2.scala
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTableH2.scala:30]: Done

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala:26]: Done

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala:85]: Done

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionIndexing.scala
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionIndexing.scala:212]: Done

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionsWriter.scala
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionsWriter.scala:53]: Done

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala
Comment (cocreature) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala:27]: I think you’re missing a crucial point here: Canton GA and DDF4O run in UCK mode and have to. NON-UCK mode has zero data continuity guarantees since we have not specd multi-key semantics in any way that we can keep stable in speedy and you fall over if they change.

Comment (daravep) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala:27]: @cocreature there are practical and theoretical issues. we'll have to pin down the semantics pretty soon once we following up with the multi-domain MVP GA release, but that is nothing that we need to do in Q2. however, the issue that we face right now is that this change is going to break quite a few of our existing tests. so the above path is to deal with these breaks. and data continuity is something we didn't promise anyway as of today.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionIndexing.scala
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionIndexing.scala]: Done

**Comment by rautenrieth-da** [comment]
Not necessary for this PR, but I think it would be nice if this situation was covered in the JdbcLedgerDao test suite. These tests use the low-level interface to access the index database, and can simulate things like state updates coming from a privacy-aware distributed ledger.

[JdbcLedgerDaoContractsSpec.scala](https://github.com/digital-asset/daml/blob/main/ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcLedgerDaoContractsSpec.scala#L237) would be a good place to add such a test.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala
Comment (oliverse-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractsTable.scala:35]: Canton follow-up tracked in https://github.com/DACH-NY/canton/issues/6493

**Pull request merged**
